components/brave_shields/browser/ad_block_service.cc

```diff
@@ -154,6 +156,21 @@ base::Optional<base::Value> AdBlockService::HiddenClassIdSelectors(
     hide_selectors = std::move(regional_selectors);
   }
 
+#if !defined(OS_ANDROID) && !defined(CHROME_OS)
+  if (!base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveAdblockCosmeticFilteringNative)) {
+    auto result_list = std::make_unique<base::ListValue>();
+    if (hide_selectors && hide_selectors->is_list()) {
+      result_list->Append(std::move(*hide_selectors));
+    }
+    if (custom_selectors && custom_selectors->is_list()) {
+      result_list->Append(std::move(*custom_selectors));
+    }
+
+    return std::move(*result_list);
+  }
+#endif
+
```

Comment:
@serg This shouldn't need to be a separate implementation between android/desktop; and it looks like the logic here is a little bit tangled up with the rest (i.e. we might be using `hide-selectors` twice)

Reply:
we will revisit later, this should be equivalent to previous logic

Reply:
the flag kBraveAdblockCosmeticFilteringNative exist on desktop only. And yes I will work in a diff PR to make the logic common as adblock_service was modified during cosmetic filter implementation and extension used a diff output comparing to a native implementation.

Reply:
Just to follow up here as I discovered a big source of confusion around this code:

In the cosmetic filtering extension version, `AdBlockService::HiddenClassIdSelectors` actually is returning _two_ lists, and is doing so by design. The return value is `[selectors, force_hide_selectors]` to comply with the `braveShields.hiddenClassIdSelectors` JS API, whose callback takes those two as arguments. This is why `custom_selectors` is appended to the list rather than having each element appended individually - it _is_ the `force_hide_selectors` array.

This is particularly confusing because `AdBlockService::HiddenClassIdSelectors` implements the same interface as `AdBlockBaseService::HiddenClassIdSelectors`, which otherwise only wraps single engines which return a single top-level list of selectors. Unfortunately `base::Value` is a pretty opaque type and so the difference in hierarchy doesn't get caught by the type checker.

This logic was previously handled within the JS binding method directly ([`BraveShieldsHiddenClassIdSelectorsFunction::Run`](https://github.com/brave/brave-core/blob/master/browser/extensions/api/brave_shields_api.cc#L93)), but I believe this detail was lost when moving the cross-engine combination logic from there to inside of `AdBlockService::HiddenClassIdSelectors`.

I think we ought to restructure the code here to reflect this. Some options; I'm currently favoring the second one:
- Modify the method signature here to return a pair of lists. We'd have to remove the `AdBlockBaseService` interface from `AdBlockService`.
- Pull the code dealing with `force_hide_selectors` into a separate method specific to `AdBlockService`, and have it only return the result from the custom filters. The two lists can then be combined for the JS API in `BraveShieldsHiddenClassIdSelectorsFunction::Run` (or the equivalent native version).

@SergeyZhukovsky have you opened any followup issue or PR for removing the platform-specific gating here? I suspect the native implementation needs to be updated to be updated accordingly as well; it looks like the return value outside of the `ifdef`s here is actually a flat list of selectors with another list at the end, i.e. `[...selectors, [...force_hide_selectors]]`.

Reply:
@antonok-edm no I haven't created any follow up yet and planned to work on that when we have a solution to use native implementation for both platforms. I was planned to just modify native method to use what the extension approach uses currently.